File: Part 8/NTSA_timefreq_instFreq.m
%%
%     COURSE: Solved challenges in neural time series analysis
%    SECTION: Time-frequency analyses
%      VIDEO: Instantaneous frequency in simulated and in real data
% Instructor: sincxpress.com
%
%%

% see also: http://mikexcohen.com/data/Cohen2014_freqslide.pdf

% simulation details
srate = 1000;
time  = (0:4*srate-1)/srate;
pnts  = length(time);
noise = 0; % noise amplitude


% create signal (multipolar chirp)
k = 10; % poles for frequencies
freqmod = 20*interp1(rand(1,k),linspace(1,k,pnts),'pchip');
signal  = sin( 2*pi * ((time + cumsum(freqmod))/srate) );
signal  = signal + randn(size(signal))*noise;


% compute instantaneous frequency
angels = angle(hilbert(signal));
instfreq = diff(unwrap(angels))/(2*pi/srate);


% static power spectrum
hz  = linspace(0,srate/2,floor(pnts/2)+1);
amp = 2*abs(fft(signal)/pnts);
amp = amp(1:length(hz)).^2;


%% plotting

figure(1), clf

% signal
subplot(311)
plot(time,signal,'k','linew',2)
xlabel('Time (s)'), ylabel('Amplitude')
title('Time domain')

% static power spectrum
subplot(312)
plot(hz,amp,'ks-','linew',2,'markerfacecolor','w')
title('Frequency domain')
set(gca,'xlim',[0 max(freqmod)*1.2])
xlabel('Frequency (Hz)'), ylabel('Amplitude')

% IF
subplot(313), hold on
plot(time(1:end-1),instfreq,'k','linew',2)
plot(time,freqmod,'r','linew',2)
set(gca,'ylim',[0 max(freqmod)*1.2])
xlabel('Time (s)'), ylabel('Frequency (Hz)')
legend({'Estimated';'Ground truth'})
title('Instantaneous frequency')

%% clean the IF: mean-smoothing

% mean smoothing kernel
k = 13;
kernel = ones(1,k)/k;
nfft = pnts + k - 1;
instfreqSmooth = ifft( fft(kernel,nfft).*fft(instfreq,nfft) );
instfreqSmooth = instfreqSmooth(ceil(k/2):end-ceil(k/2));

plot(time(1:end-1),instfreqSmooth,'b','linew',3)

%% in real data!

load EEGrestingState.mat
time = (0:length(eegdata)-1)/srate;

figure(2), clf
pwelch(eegdata,srate,srate/2,srate*2,srate)
set(gca,'xlim',[0 40])


%% narrowband filter around 10 Hz (FIR)

% filter parameters
nyquist = srate/2;
frange  = [8 12];
order   = round( 20*srate/frange(1) );

% filter kernel
filtkern = fir1(order,frange/nyquist);

% compute the power spectrum of the filter kernel
filtpow = abs(fft(filtkern)).^2;
% compute the frequencies vector and remove negative frequencies
hz      = linspace(0,srate/2,floor(length(filtkern)/2)+1);
filtpow = filtpow(1:length(hz));


%%% visualize the filter kernel
figure(3), clf
subplot(121)
plot(filtkern,'linew',2)
xlabel('Time points')
title('Filter kernel (fir1)')
axis square

% plot amplitude spectrum of the filter kernel
subplot(122), hold on
plot(hz,filtpow,'ks-','linew',2,'markerfacecolor','w')
plot([0 frange(1) frange frange(2) nyquist],[0 0 1 1 0 0],'ro-','linew',2,'markerfacecolor','w')

% dotted line corresponding to the lower edge of the filter cut-off
plot([1 1]*frange(1),get(gca,'ylim'),'k:')

% make the plot look nicer
set(gca,'xlim',[0 frange(1)*4])%,'ylim',[-.05 1.05])
xlabel('Frequency (Hz)'), ylabel('Filter gain')
legend({'Actual';'Ideal'})
title('Frequency response of filter (fir1)')
axis square


%% IF on filtered data

% apply the filter to the data
feegdata = filtfilt(filtkern,1,double(eegdata));

% plot the data for comparison
figure(4), clf
plot(time,eegdata, time,feegdata ,'linew',2)
xlabel('Time (s)'), ylabel('Amplitude')
zoom on



% compute instantaneous frequency
angels = angle(hilbert(feegdata));
instalpha = diff(unwrap(angels))/(2*pi/srate);

figure(5), clf
plot(time(1:end-1),instalpha,'ks-','markerfacecolor','w')
xlabel('Time (s)'), ylabel('Frequency (Hz)')
zoom on

%% apply median filter to supra-threshold points

% convert to z and show histogram
instz = (instalpha-mean(instalpha)) / std(instalpha);

figure(6), clf
hist(instz,200)
xlabel('I.F. (z)'), ylabel('Count')
title('Distribution of z-norm. IF')

% identify supra-threshold data points
tofilter = find( abs(instz)>2 );

% now for median filter
instalphaFilt = instalpha;
k = round(1000*50/srate); % median kernel size is 2k+1, where k is time in ms
for i=1:length(tofilter)
    indices = max(1,tofilter(i)-k):min(pnts,tofilter(i)+k);
    instalphaFilt(tofilter(i)) = median(instalpha(indices));
end

figure(7), clf, hold on
plot(time(1:end-1),instalpha,'k','linew',2)
plot(time(1:end-1),instalphaFilt,'r--','linew',2)
xlabel('Time (s)'), ylabel('Frequency (Hz)')
legend({'Original';'Filtered'})
set(gca,'ylim',[5 15])
zoom on

%% done.

File: Part 8/NTSA_timefreq_instFreq.py
# %%
#     COURSE: Solved problems in neural time series analysis
#    SECTION: Time-frequency analyses
#    TEACHER: Mike X Cohen, sincxpress.com
#      VIDEO: Visualize time-frequency power from all channels

# !%matplotlib inline
import os
import sys

import matplotlib.pyplot as plt
import numpy as np
from numpy.fft import fft, ifft
from scipy.io import loadmat
from scipy.signal import filtfilt, firwin, hilbert, welch

# Set figure settings
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
from utils import set_fig_dpi, set_style

set_fig_dpi(), set_style()

# %%
# Simulation details
srate = 1000
time = np.arange(0, 4, 1 / srate)
pnts = len(time)
noise_amplitude = 0.1  # Noise amplitude --> if non-zero, smoothing must be applied

# Create signal (multipolar chirp)
k = 10  # Poles for frequencies
freq_mod = 20 * np.interp(np.linspace(0, k, pnts), np.arange(0, k), np.random.rand(k))
signal = np.sin(2 * np.pi * (time + np.cumsum(freq_mod) / srate))
signal += np.random.randn(len(signal)) * noise_amplitude

# Compute instantaneous frequency
signal_hilbert = hilbert(signal)
angles = np.angle(signal_hilbert)
inst_freq = np.diff(np.unwrap(angles)) / (2 * np.pi) * srate

# Static power spectrum
hz = np.linspace(0, srate / 2, int(np.floor(pnts / 2)) + 1)
amp = 2 * np.abs(np.fft.fft(signal) / pnts)
amp = amp[: len(hz)] ** 2

# Mean smoothing kernel
k = 13
kernel = np.ones(k) / k
nfft = len(inst_freq) + k - 1

# Apply mean smoothing
inst_freq_fft = fft(inst_freq, nfft)
kernel_fft = fft(kernel, nfft)
inst_freq_smooth = ifft(inst_freq_fft * kernel_fft)
inst_freq_smooth = inst_freq_smooth[int(np.ceil(k / 2)) - 1 : -int(np.ceil(k / 2))]


# Plotting
plt.figure()

# Time domain signal
plt.subplot(311)
plt.plot(time, signal, 'k')
plt.xlabel('Time (s)')
plt.ylabel('Amplitude')
plt.title('Time Domain')

# Frequency domain
plt.subplot(312)
plt.plot(hz, amp, 'ks-', markerfacecolor='w')
plt.xlim([0, np.max(freq_mod) * 1.2])
plt.xlabel('Frequency (Hz)')
plt.ylabel('Power')
plt.title('Frequency Domain')

# Instantaneous frequency
plt.subplot(313)
# plt.plot(time[:-1], inst_freq, 'k', label='Estimated')  # Non-smoothened
plt.plot(time[1:-1], inst_freq_smooth, 'k', label='Estimated')  # Smoothened
plt.plot(time, freq_mod, 'r', label='Ground Truth')
plt.ylim([0, np.max(freq_mod) * 1.2])
plt.xlabel('Time (s)')
plt.ylabel('Frequency (Hz)')
plt.title('Instantaneous Frequency')
plt.legend()

plt.tight_layout()
plt.show()

# %% Now for real data
# Load the data
mat_file = loadmat(os.path.join('..', 'data', 'EEGrestingState.mat'))
data = mat_file['eegdata'][0]
srate = mat_file['srate'][0][0]
n_samples = len(data)

# Get the time vector
time = np.arange(0, n_samples) / srate

frequencies, psd = welch(
    data, fs=srate, nperseg=srate, noverlap=srate // 2, nfft=srate * 2
)
psd = 10 * np.log10(psd)  # Convert to dB

plt.figure()
plt.plot(frequencies, psd)
plt.xlabel('Frequency (Hz)')
plt.ylabel('Power')
plt.title('Welch Power Spectral Density Estimate')
plt.xlim(0, 40)
plt.ylim(-15, 15)
plt.show()

# %% Narrowband filter around 10 Hz (alpha band)
# Filter parameters
nyquist = srate / 2
freq_range = np.array([8, 12])
order = round(20 * srate / freq_range[0])
order |= 1  # Ensure it is odd

# Filter kernel
filt_kern = firwin(order + 1, freq_range, nyq=nyquist, pass_zero=False)

# Compute the power spectrum of the filter kernel
filt_pow = np.abs(np.fft.fft(filt_kern)) ** 2
# Compute the frequencies vector and remove negative frequencies
hz = np.linspace(0, srate / 2, int(np.floor(len(filt_kern) / 2)) + 1)
filt_pow = filt_pow[: len(hz)]

# Visualize the filter kernel
plt.figure()
plt.subplot(121)
plt.plot(filt_kern)
plt.xlabel('Time Points')
plt.title('Filter Kernel (firwin)')

# Plot amplitude spectrum of the filter kernel
plt.subplot(122)
plt.plot(hz, filt_pow, 'ks-', markerfacecolor='w', label='actual')
plt.plot(
    [0, freq_range[0], freq_range[0], freq_range[1], freq_range[1], nyquist],
    [0, 0, 1, 1, 0, 0],
    'ro-',
    markerfacecolor='w',
    label='ideal',
)
plt.plot([freq_range[0], freq_range[0]], plt.gca().get_ylim(), 'k:')
plt.xlim(0, freq_range[0] * 4)
plt.xlabel('Frequency (Hz)')
plt.ylabel('Filter gain')
plt.legend()
plt.title('Frequency Response of Filter (firwin)')

plt.tight_layout()
plt.show()

# %%
# Apply the filter to the data
filt_data = filtfilt(filt_kern, 1, data.astype(float))

# Plot the data for comparison
plt.figure()
plt.plot(time, data, label='Original')
plt.plot(time, filt_data, label='Filtered')
plt.xlabel('Time (s)')
plt.ylabel('Amplitude')
plt.legend()
plt.show()

# %% Compute instantaneous frequency
angles = np.angle(hilbert(filt_data))
inst_alpha = np.diff(np.unwrap(angles)) / (2 * np.pi / srate)

plt.figure()
plt.plot(time[:-1], inst_alpha, 'ks-', markerfacecolor='w')
plt.xlabel('Time (s)')
plt.ylabel('Frequency (Hz)')
plt.show()

# %% Apply median filter to supra-threshold points

# Convert to z-score and show histogram
inst_z = (inst_alpha - np.mean(inst_alpha)) / np.std(inst_alpha)

plt.figure()
plt.hist(inst_z, 200)
plt.xlabel('I.F. (z)'), plt.ylabel('Count')
plt.title('Distribution of z-normalized Instantaneous Frequency')
plt.show()

# %% Cut off large deviations using median filter

# Identify supra-threshold data points
# 2 standard deviations corresponds to 95% confidence interval
to_filter = np.where(np.abs(inst_z) > 2)[0]

# Now for median filter
inst_alpha_filt = inst_alpha.copy()
k = round(1000 * 50 / srate)  # Median kernel size is 2k+1, where k is time in ms
for i in to_filter:
    indices = np.arange(max(0, i - k), min(len(inst_alpha), i + k + 1))
    inst_alpha_filt[i] = np.median(inst_alpha[indices])

plt.figure()
plt.plot(time[:-1], inst_alpha, 'k', label='Original')
plt.plot(time[:-1], inst_alpha_filt, 'r--', label='Filtered')
plt.xlabel('Time (s)')
plt.ylabel('Frequency (Hz)')
plt.legend()
plt.ylim([5, 15])
plt.show()

# %%
